File: js/engine/collision.js
// Rectangle-circle and circle-circle collision math
// PURE — no DOM, no browser APIs

// Find nearest point on axis-aligned rectangle to point (px, py)
// Rectangle defined by center (cx, cy) and half-widths (hw, hh)
export function nearestPointOnRect(px, py, cx, cy, hw, hh) {
  const nx = Math.max(cx - hw, Math.min(px, cx + hw));
  const ny = Math.max(cy - hh, Math.min(py, cy + hh));
  return { x: nx, y: ny };
}

// Check if circle (bx, by, radius) overlaps rectangle (cx, cy, hw, hh)
export function circleRectOverlap(bx, by, radius, cx, cy, hw, hh) {
  const nearest = nearestPointOnRect(bx, by, cx, cy, hw, hh);
  const dx = bx - nearest.x;
  const dy = by - nearest.y;
  return (dx * dx + dy * dy) < (radius * radius);
}

// Determine which face of rectangle the circle penetrated least
// Returns 'left', 'right', 'top', or 'bottom'
export function minOverlapFace(bx, by, radius, cx, cy, hw, hh) {
  const overlapLeft   = (bx + radius) - (cx - hw);
  const overlapRight  = (cx + hw) - (bx - radius);
  const overlapTop    = (by + radius) - (cy - hh);
  const overlapBottom = (cy + hh) - (by - radius);

  const min = Math.min(overlapLeft, overlapRight, overlapTop, overlapBottom);

  if (min === overlapLeft)   return 'left';
  if (min === overlapRight)  return 'right';
  if (min === overlapTop)    return 'top';
  return 'bottom';
}

// Check if two circles overlap
export function circleCircleOverlap(x1, y1, r1, x2, y2, r2) {
  const dx = x2 - x1;
  const dy = y2 - y1;
  const dist = Math.sqrt(dx * dx + dy * dy);
  const minDist = r1 + r2;
  return { overlapping: dist < minDist, dist, dx, dy, minDist };
}

File: plugin/Source/Engine/Collision.cpp
#include "Collision.h"

namespace Collision
{

bool circleRectOverlap(float bx, float by, float radius,
                       float cx, float cy, float hw, float hh)
{
    float nearestX = std::clamp(bx, cx - hw, cx + hw);
    float nearestY = std::clamp(by, cy - hh, cy + hh);
    float dx = bx - nearestX;
    float dy = by - nearestY;
    return (dx * dx + dy * dy) < (radius * radius);
}

RectFace minOverlapFace(float bx, float by, float radius,
                        float cx, float cy, float hw, float hh)
{
    float overlapLeft  = (bx + radius) - (cx - hw);
    float overlapRight = (cx + hw) - (bx - radius);
    float overlapTop   = (by + radius) - (cy - hh);
    float overlapBot   = (cy + hh) - (by - radius);

    float minVal = overlapLeft;
    RectFace face = RectFace::Left;

    if (overlapRight < minVal) { minVal = overlapRight; face = RectFace::Right; }
    if (overlapTop < minVal)   { minVal = overlapTop;   face = RectFace::Top; }
    if (overlapBot < minVal)   { face = RectFace::Bottom; }

    return face;
}

CircleCircleResult circleCircleOverlap(float x1, float y1, float r1,
                                       float x2, float y2, float r2)
{
    CircleCircleResult result;
    result.dx = x2 - x1;
    result.dy = y2 - y1;
    result.dist = std::sqrt(result.dx * result.dx + result.dy * result.dy);
    result.minDist = r1 + r2;
    result.overlapping = result.dist < result.minDist;
    return result;
}

} // namespace Collision
